File: api/app/schemas/article.py
import uuid
from datetime import datetime
from typing import Annotated, Optional

from pydantic import Field, HttpUrl

from app.schemas import AppBaseModel


class ArticleCreate(AppBaseModel):
    """POST /articles request body (called from n8n).

    published_at is received as ISO8601 UTC datetime (e.g. "2026-01-01T00:00:00Z").
    n8n converts the /ingest published_at string to datetime before sending.
    """

    source_url: HttpUrl
    source_name: Optional[str] = Field(None, max_length=100)
    title_original: Optional[str] = Field(None, max_length=500)
    title_ja: Optional[str] = Field(None, max_length=500)
    body_original: Optional[str] = Field(None, max_length=200000)
    body_translated: Optional[str] = Field(None, max_length=200000)
    summary_ja: Optional[str] = Field(None, max_length=5000)
    author: Optional[str] = Field(None, max_length=200)
    published_at: Optional[datetime] = None
    og_image_url: Optional[str] = Field(None, max_length=2083)
    categories: Optional[list[Annotated[str, Field(max_length=50)]]] = Field(None, max_length=20)
    metadata: Optional[dict] = None


class ArticleListItem(AppBaseModel):
    """GET /articles response item (summary only)."""

    id: uuid.UUID
    source_url: str
    source_name: Optional[str] = None
    title_ja: Optional[str] = None
    summary_ja: Optional[str] = None
    author: Optional[str] = None
    published_at: Optional[datetime] = None
    og_image_url: Optional[str] = None
    categories: Optional[list[str]] = None
    created_at: datetime


class ArticleDetail(AppBaseModel):
    """GET /articles/{id} response (detail with translated body)."""

    id: uuid.UUID
    source_url: str
    source_name: Optional[str] = None
    title_original: Optional[str] = None
    title_ja: Optional[str] = None
    body_translated: Optional[str] = None
    summary_ja: Optional[str] = None
    author: Optional[str] = None
    published_at: Optional[datetime] = None
    og_image_url: Optional[str] = None
    categories: Optional[list[str]] = None
    metadata: Optional[dict] = Field(None, validation_alias="metadata_")
    created_at: datetime


class ArticleListResponse(AppBaseModel):
    """GET /articles paginated response."""

    items: list[ArticleListItem]
    total: int
    page: int
    per_page: int


class ArticleCheckResponse(AppBaseModel):
    """GET /articles/check response."""

    exists: bool

File: api/src/schemas/article.ts
import { z } from "zod";
import { dateString, isoDateTime, paginatedSchema, paginationQuery } from "./base.js";

export const articleCreateSchema = z
  .object({
    source_url: z.string().url(),
    source_name: z.string().max(100).nullish(),
    title_original: z.string().max(500).nullish(),
    title_ja: z.string().max(500).nullish(),
    body_original: z.string().max(200000).nullish(),
    body_translated: z.string().max(200000).nullish(),
    summary_ja: z.string().max(5000).nullish(),
    author: z.string().max(200).nullish(),
    published_at: z.string().datetime({ offset: true }).nullish(),
    og_image_url: z.string().url().max(2083).nullish(),
    categories: z.array(z.string().max(50)).max(20).nullish(),
    metadata: z.record(z.unknown()).nullish(),
  })
  .strict();

export type ArticleCreate = z.infer<typeof articleCreateSchema>;

export const articleListQuerySchema = paginationQuery.extend({
  date: dateString.optional(),
  category: z.string().max(50).optional(),
});

export const articleCheckQuerySchema = z.object({
  url: z.string().url(),
});

export const articleCheckResponseSchema = z.object({
  exists: z.boolean(),
});

export const articleListItemSchema = z.object({
  id: z.string().uuid(),
  source_url: z.string(),
  source_name: z.string().nullable(),
  title_ja: z.string().nullable(),
  summary_ja: z.string().nullable(),
  author: z.string().nullable(),
  published_at: isoDateTime.nullable(),
  og_image_url: z.string().nullable(),
  categories: z.array(z.string()).nullable(),
  created_at: isoDateTime,
});

export type ArticleListItem = z.infer<typeof articleListItemSchema>;

export const articleListResponseSchema = paginatedSchema(articleListItemSchema);

export const articleDetailSchema = articleListItemSchema.extend({
  title_original: z.string().nullable(),
  body_original: z.string().nullable(),
  body_translated: z.string().nullable(),
  metadata: z.record(z.unknown()).nullable(),
});

export type ArticleDetail = z.infer<typeof articleDetailSchema>;

export const articleNeighborItemSchema = z.object({
  id: z.string().uuid(),
  title_ja: z.string().nullable(),
  og_image_url: z.string().nullable(),
  published_at: isoDateTime.nullable(),
});

export type ArticleNeighborItem = z.infer<typeof articleNeighborItemSchema>;

export const articleNeighborsResponseSchema = z.object({
  prev: articleNeighborItemSchema.nullable(),
  next: articleNeighborItemSchema.nullable(),
});

export type ArticleNeighborsResponse = z.infer<typeof articleNeighborsResponseSchema>;
